proxy/src/metrics.rs

```diff
@@ -20,159 +29,228 @@ const DEFAULT_HTTP_REPORTING_TIMEOUT: Duration = Duration::from_secs(60);
         let res = collect_metrics_iteration(
             &http_client,
-            &mut cached_metrics,
+            &mut proxy_counter_queue,
+            &mut events,
             &config.endpoint,
             &hostname,
+            report_delta,
         )
         .await;
 
         match res {
             Err(e) => error!("failed to send consumption metrics: {e} "),
-            Ok(_) => trace!("periodic metrics collection completed successfully"),
+            Ok(_) => {
+                trace!("periodic metrics collection completed successfully")
+            }
         }
     }
 }
 
-fn gather_proxy_io_bytes_per_client() -> Vec<(Ids, (u64, DateTime<Utc>))> {
-    let mut current_metrics: Vec<(Ids, (u64, DateTime<Utc>))> = Vec::new();
-    let metrics = prometheus::default_registry().gather();
-
-    for m in metrics {
-        if m.get_name() == "proxy_io_bytes_per_client" {
-            for ms in m.get_metric() {
-                let direction = ms
-                    .get_label()
-                    .iter()
-                    .find(|l| l.get_name() == "direction")
-                    .unwrap()
-                    .get_value();
-
-                // Only collect metric for outbound traffic
-                if direction == "tx" {
-                    let endpoint_id = ms
-                        .get_label()
-                        .iter()
-                        .find(|l| l.get_name() == "endpoint_id")
-                        .unwrap()
-                        .get_value();
-                    let branch_id = ms
-                        .get_label()
-                        .iter()
-                        .find(|l| l.get_name() == "branch_id")
-                        .unwrap()
-                        .get_value();
-
-                    let value = ms.get_counter().get_value() as u64;
-
-                    // Report if the metric value is suspiciously large
-                    if value > (1u64 << 40) {
-                        warn!(
-                            "potentially abnormal counter value: branch_id {} endpoint_id {} val: {}",
-                            branch_id, endpoint_id, value
-                        );
-                    }
+/// Counter for proxy bytes transferred
+pub struct ProxyCounter {
+    ids: Ids,
+    /// Outbound bytes from neon compute to client
+    pub tx: AtomicU64,
+}
 
-                    current_metrics.push((
-                        Ids {
-                            endpoint_id: endpoint_id.to_string(),
-                            branch_id: branch_id.to_string(),
-                        },
-                        (value, Utc::now()),
-                    ));
-                }
+pub struct CounterState {
+    counter: Arc<ProxyCounter>,
+    last_seen: DateTime<Utc>,
+    total_tx: u64,
+}
+
+impl CounterState {
+    fn is_closed(&self) -> bool {
+        // we are the only holders of this counter
+        Arc::strong_count(&self.counter) == 1
```

Comment:
I wonder if we should use `Arc::get_mut(&mut self.counter).is_some()` instead because the impl of that is not just a counter read https://doc.rust-lang.org/stable/src/alloc/sync.rs.html#1807-1828.

Maybe similarly for the re-queueing.